### PR TITLE
feat: balance tx fails fast on txcoll or collRet

### PR DESCRIPTION
**Pull request by Unisay**
`balanceTransaction` fails gracefully if Tx body contains either total collateral attribute or collateral return outputs attribute.

### Issue Number

ADP-1653

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: Perhaps in `ErrBalanceTxExistingReturnCollateral ` in particular, it might be worth tweaking the wording to hint that we might support it in the future.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: I think we have come to use `CreatedInvalidTransaction` everywhere in `IsServerError ErrBalanceTx` from the beginning just by accident. It _might_ be better to break the pattern now and create new `ApiErrorCode` constructors.

File: lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs
Comment (Anviking) [inline lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3419]: 👍

**Review by Anviking — COMMENTED** [review]
Neat!

For testing I think it would be very nice if we could:

1. make [`prop_balanceTransactionValid`](https://github.com/input-output-hk/cardano-wallet/blob/f5cff1c23260c6985f9007b0b5ac8fb29be631ae/lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs#L3224) run in both `Alonzo` and `Babbage` which would be valuable irrespective of this task. I think and hope this is not that complicated.

2. Generate a tiny amount of totalCollateral and returnCollateral in https://github.com/input-output-hk/cardano-wallet/blob/f5cff1c23260c6985f9007b0b5ac8fb29be631ae/lib/core/src/Cardano/Api/Gen.hs#L1335-L1337

And actually, we could then perhaps add some checks [here](https://github.com/input-output-hk/cardano-wallet/blob/f5cff1c23260c6985f9007b0b5ac8fb29be631ae/lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs#L3249-L3263) to mimic the ledger rules

Skärmavbild 2022-06-20 kl  12 47 33


to make the tests fail unless we have `guardExisting{Return, Total}Collateral `.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: There is https://input-output.atlassian.net/browse/ADP-1684, but perhaps allowing pre-set `collRet` is not on the table 🤔  - never mind then.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Unisay) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: 🆗

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: Suggestion: add a hyphen to make it clear that "zero-ada" is an adjective of "outputs":

```suggestion
                , "one or more zero-ada outputs. In the future I might be able"
```

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]:
```suggestion
            apiError err500 BalanceTxUnderestimatedFee $ unwords
                [ "I have somehow underestimated the fee of the transaction by"
                , pretty c
                , "and cannot finish balancing."
                ]
```

File: lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs
Comment (jonathanknowles) [inline lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3417]: @Unisay I think these labels are great, thanks for adding them.

However, I think we still have one issue with this approach, namely that using `label` within the body of a pattern match doesn't actually cause QuickCheck to verify that this particular case is covered during a test run.

This could mean that some of the patterns are not matched in any given run. We can of course manually inspect the test output to verify that all labels are there, but it's likely that people will forget. If a particular pattern is _never_ matched, this could result in a false sense of security that the implementation is correct (and might mean that we fail to spot a regression later on).

For an automated coverage check, I think we'd need to use some combination of `checkCoverage` and `cover` _outside_ of the pattern match. See:

- https://input-output.atlassian.net/browse/ADP-1843
- https://github.com/input-output-hk/cardano-wallet/pull/3255#discussion_r886358943

It shouldn't block this PR from being merged (because this PR is a strict improvement over `master`), but I do think we should add coverage checks in a future PR.

**Review by jonathanknowles — COMMENTED** [review]
Hi @Unisay 

Thanks for making this PR! I've left a few suggestions and comments.

File: lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs
Comment (Unisay) [reply lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3417]: I completely agree that coverage checks should be implemented. I also agree that its a good idea not to block this PR. https://input-output.atlassian.net/browse/ADP-1843 seems like a proper task for it.

File: specifications/api/swagger.yaml
Comment (Anviking) [inline specifications/api/swagger.yaml:4573]: 👌 🙏

File: specifications/api/swagger.yaml
Comment (Anviking) [inline specifications/api/swagger.yaml]: I presume most changes in `swagger.yaml` are reformatting? Could you please pull it out into a separate commit both to make it easier to review the actual changes and for history?

Comment (Anviking) [reply specifications/api/swagger.yaml]: Same thing, but to a lesser extent with the changes to existing error codes in `ErrBalanceTx`: I think those could fairly easily be pulled out into a preceding preparatory commit (keeping the addition of the new errors in the existing, main commit)

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: For the record: only meant to suggest to add the new errors in a different style, but changing the existing errors too is even better - thanks!

**Review by Anviking — COMMENTED** [review]
Thanks a lot, LGTM! As discussed it would be cool to extend `prop_balanceTransactionValid` with the collateral rules for a "higher level" property, but we don't have to do that now.

But I would like ask to split up at least the swagger-reformatting into a separate commit before merging (see comments).

**Review by Anviking — APPROVED** [review]

File: specifications/api/swagger.yaml
Comment (Unisay) [reply specifications/api/swagger.yaml]: That formatting was done inadvertently, reverted.

**Pull request merged**
